File: common/src/main/java/de/nicidienase/chaosflix/common/entities/recording/Metadata.java
package de.nicidienase.chaosflix.common.entities.recording;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;
import com.orm.SugarRecord;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by felix on 06.04.17.
 */

public class Metadata extends SugarRecord implements Parcelable {
	Map<Long,Long> related;
	@SerializedName("remote_id")
	String remoteId;

	protected Metadata(Parcel in) {
		remoteId = in.readString();
		long mapSize = in.readLong();
		related = new HashMap<>();
		for(int i= 0; i< mapSize; i++){
			related.put(in.readLong(),in.readLong());
		}
	}

	public static final Creator<Metadata> CREATOR = new Creator<Metadata>() {
		@Override
		public Metadata createFromParcel(Parcel in) {
			return new Metadata(in);
		}

		@Override
		public Metadata[] newArray(int size) {
			return new Metadata[size];
		}
	};

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeString(remoteId);
		dest.writeLong(related.size());
		for(Map.Entry<Long,Long> entry: related.entrySet()){
			dest.writeLong(entry.getKey());
			dest.writeLong(entry.getValue());
		}
	}

	public Map<Long, Long> getRelated() {
		return related;
	}

	public void setRelated(Map<Long, Long> related) {
		this.related = related;
	}

	public String getRemoteId() {
		return remoteId;
	}

	public void setRemoteId(String remoteId) {
		this.remoteId = remoteId;
	}
}

File: common/src/main/java/de/nicidienase/chaosflix/common/entities/recording/Metadata.kt
package de.nicidienase.chaosflix.common.entities.recording

import android.os.Parcel
import android.os.Parcelable
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import de.nicidienase.chaosflix.common.entities.Converters

@JsonIgnoreProperties(ignoreUnknown = true)
open class Metadata(
        var related: Map<Long,Long>?,
        @JsonProperty("remote_id") var remoteId: String?
) : Parcelable {
    constructor(parcel: Parcel) : this(null,parcel.readString()) {
        val mapSize = parcel.readInt()
        val stringRepresentation = parcel.readString()
        if(stringRepresentation.isNotEmpty()){
            related = Converters().stringToLongMap(stringRepresentation)
        } else {
            related = HashMap<Long,Long>()
        }
//        for (i in 0..mapSize){
//            map.put(parcel.readLong(),parcel.readLong())
//        }
//        related = map
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(remoteId)
        if(related!=null){
            related?.let {
                parcel.writeString(Converters().longMapToString(it))
            }
        } else {
            parcel.writeString("")
        }
//        related?.run {
//            parcel.writeInt(this.size)
//            this.map { entry -> parcel.writeLong(entry.key); parcel.writeLong(entry.value) }
//        }
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Metadata> {
        override fun createFromParcel(parcel: Parcel): Metadata {
            return Metadata(parcel)
        }

        override fun newArray(size: Int): Array<Metadata?> {
            return arrayOfNulls(size)
        }
    }
}
